### ISSUE TITLE
Deprecate torch.(min|max|median|mode) to only return values to be more compatible with Python Array API Standard

### ISSUE THREAD

**Issue by heitorschueroff** [pytorch/pytorch#61490]
`torch.min`, `torch.max`, `torch.median` and `torch.mode` return indices along with the values when reducing over one dimension. This is incompatible with the Python Array API Standard as described in the following issue https://github.com/pytorch/pytorch/issues/58745 and furthermore makes it difficult to support reducing over multiple dimensions. Not to mention that the same operator having multiple return values is not a good UX and if the user does not need the indices we can achieve better performance by only computing the values which allows for better vectorization.

I propose we deprecate these operators to only return values by default and allow reducing over multiple dimensions, and if needed we can introduce a keyword argument `return_indices=False` to return the indices or implement `torch.argmedian` and `torch.argmode` similar to `torch.argmax`.

**Comment by mruberry** [comment]
We should think how to best understand the community's desire for consistency with NumPy vs. the cost of changing these operators' behavior.

**Comment by heitorschueroff** [comment]
After some thought I think an API where optional `out`-like tensors can be given to collect extra output information is probably the best approach for the following reasons:
- Users don't have to remember the names of all possible variants, e.g. `max` and `max_with_indices`
- We avoid creating a combinatorial number of variants in cases like `torch.unique`
- The output type does not change as these extra 'out'-like tensors are not returned as is the case with the current`out` parameter.
- We already have support for `out=` parameter

**Example API**
```
max(input, *, dim=None, keepdim=False, out=None, indices=None) -> Tensor
```

@albanD How do you think this would work with autograd?

**Comment by albanD** [comment]
out= arguments don't work with autograd no. And, while we could, that would be a lot of work and there is no plan to do it anytime soon.

Another thing I would say: the gradient when indices are returned or not are not the same. So changing the "do we return indices" also implies modifying the gradient formula.

**Comment by mruberry** [comment]
> Another thing I would say: the gradient when indices are returned or not are not the same. So changing the "do we return indices" also implies modifying the gradient formula.

We could also compute the indices and not return them to the user (a pattern we use for some other ops)

**Comment by albanD** [comment]
The problem is not necessarily with the fact that we have the indices or not in the backend.
It is more that the mathematical function we expose to the user is different as in one case, there is a single max value (as specified by the returned index) while in the other case, any input with the maximum value is considered.
So if the function stop returning indices to the user we need to change the gradient that the user gets.

**Comment by ngimel** [comment]
What is the proposed alternative for current `max|min` functions that preserves the behavior?

**Comment by rgommers** [comment]
> It is more that the mathematical function we expose to the user is different as in one case, there is a single max value (as specified by the returned index) while in the other case, any input with the maximum value is considered.
> So if the function stop returning indices to the user we need to change the gradient that the user gets.

@albanD I'm not sure I understand this. How is the gradient different for these two cases for example?
```python
>>> t = torch.tensor([1,3,1,3,1,3,0])
>>> torch.max(t)  # case 1
tensor(3)
>>> torch.argmax(t)
tensor(1)
>>> torch.max(t, dim=0)  # case 2
torch.return_types.max(
values=tensor(3),
indices=tensor(1))
```
It looks like the exact same mathematical function to me, and if the definition switches based on whether one adds `dim=0`, then that's a problem I'd think?

**Comment by albanD** [comment]
> It looks like the exact same mathematical function to me, and if the definition switches based on whether one adds dim=0, then that's a problem I'd think?

It does change between `torch.max(t)` and `torch.max(t, dim=0)`.
The problem is that `t` contains multiple times the same value. That makes the current point non-differentiable.
So we are left with sub-differentials (a whole bunch of them) and we need to pick one.

In the case where we return an index for the max, we pick the sub-differential that set all the gradients on the input matching the returned index. This is what users expect if they use that index to do further processing/filtering on the input.
When we don't return the indices, then we follow the rules we have for such function: "use the minimum-norm subgradient". In this case, it is the one that will "split equally" the gradient for every input that achieves the max value.

**Comment by heitorschueroff** [comment]
> > It looks like the exact same mathematical function to me, and if the definition switches based on whether one adds dim=0, then that's a problem I'd think?
> 
> It does change between `torch.max(t)` and `torch.max(t, dim=0)`.
> The problem is that `t` contains multiple times the same value. That makes the current point non-differentiable.
> So we are left with sub-differentials (a whole bunch of them) and we need to pick one.
> 
> In the case where we return an index for the max, we pick the sub-differential that set all the gradients on the input matching the returned index. This is what users expect if they use that index to do further processing/filtering on the input.
> When we don't return the indices, then we follow the rules we have for such function: "use the minimum-norm subgradient". In this case, it is the one that will "split equally" the gradient for every input that achieves the max value.

Thank you for the detailed description @albanD. Wouldn't it be possible then to have two signatures in native_functions.yaml, one that accepts the `*, Tensor indices` parameter and another that does not and set different gradients for each. This way users still get the benefit of being able to call `torch.max(t, indices=<some_tensor>)`.

**Comment by ngimel** [comment]
>Thank you for the detailed description @albanD. Wouldn't it be possible then to have two signatures in native_functions.yaml, one that accepts the *, Tensor indices parameter and another that does not and set different gradients for each. This way users still get the benefit of being able to call torch.max(t, indices=<some_tensor>)

How is it different from the current `max` function that we have? Seems like a lot of work, breaking a lot of users who currently use `max` to achieve the exact state that we have today.

**Comment by heitorschueroff** [comment]
> > Thank you for the detailed description @albanD. Wouldn't it be possible then to have two signatures in native_functions.yaml, one that accepts the *, Tensor indices parameter and another that does not and set different gradients for each. This way users still get the benefit of being able to call torch.max(t, indices=<some_tensor>)
> 
> How is it different from the current `max` function that we have? Seems like a lot of work, breaking a lot of users who currently use `max` to achieve the exact state that we have today.

@ngimel Those are valid concerns and we should definitely discuss them. The motivation for deprecating `torch.max` and similar operators to stop returning indices by default is so:

1. We are compatible with the Python Array API Standard (see [here](https://data-apis.org/array-api/latest/API_specification/statistical_functions.html#max-x-axis-none-keepdims-false))
2. We can compute max along multiple dimensions. This would still be possible if we returned a linearized index into the multiple dimensions and added a functions such as [`np.unravel_index`](https://numpy.org/doc/stable/reference/generated/numpy.unravel_index.html) so the user can get the multi-dimensional index if needed.
3. If users don't need the indices, we can offer a more optimized kernel.

**Comment by ngimel** [comment]
There are `amin|amax` that support tuple of dimensions and use efficient kernels as they don't need to return indices. Returning an index for a tuple of dimensions is not a very useful or intuitive operation - would linearized index be different if user passes (0,1) or (1,0)? Why or why not? 
Also, the latest proposal is to return the same type (tensor of max/min) but mutate a kwarg  tensor if requested? I don't see how it's a better UX.

**Comment by mruberry** [comment]
Before we get too deep into this I think there are a couple facts we need:

- how often do people use both return values from max
- how slow is calling max and then arg_max

Even if we decide not to deprecate max I'd be curious to see a function like `max_with_indices` (name TBD) that always returned both the maximum values and their indices to avoid max's odd polymorphism.

**Comment by ngimel** [comment]
>how often do people use both return values from max
how slow is calling max and then arg_max

The issue, as @alband pointed out, is not whether people use both return values from max, it's that even if only value is used, gradient is different compared to amax. 
Calling `max` and `argmax` is, unsurprisingly, approximately twice as slow as calling one of them. 

Also, I would like to understand what are we gaining by turning`max/min` into an alias for already existing numpy compliant `amax/amin`, other than being compliant with yet another standard that is being developed.

**Comment by rgommers** [comment]
`amax`/`amin` are weird aliases in numpy that are very infrequently used, `max`/`min` are the normal functions. The only reason `amax`/`amin` exist is because long ago it was considered a problem to alias builtins (from the time people did `from pylab import *`).

> The issue, as @albanD pointed out, is not whether people use both return values from max, it's that even if only value is used, gradient is different compared to amax.

This is true but note that cause and effect are switched. The different math follows from the polymorphism, not the other way around. And polymorphism is still bad here. The current implementation makes very little sense.

**Comment by ngimel** [comment]
Even if we reduce the space of functions we are discussing to `max(dim=..)` and `amax(dim=..)`, there's different behavior even though there's no polymorphism.
